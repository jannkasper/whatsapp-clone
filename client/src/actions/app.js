
export const receiveAuthentication = payload => async dispatch => {
    dispatch({ type: "USER_AUTHENTICATION", payload });
};

export const openContactsNavigation = payload => async dispatch => {
    dispatch({ type: "OPEN_CONTACTS_NAVIGATION", payload });
};

export const closeContactsNavigation = payload => async dispatch => {
    dispatch({ type: "CLOSE_CONTACTS_NAVIGATION", payload });
};

export const setSelectedContact = payload => async dispatch => {
    dispatch({ type: "SET_SELECTED_CONTACT", payload });
};