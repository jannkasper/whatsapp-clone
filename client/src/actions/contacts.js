export const fetchContactsPending = () => ({ type: "FETCH_CONTACTS_PENDING" });


export const fetchContactsSuccess = payload => async dispatch => {
    dispatch({ type: "FETCH_CONTACTS_SUCCESS", payload });
};

export const fetchContactsError = payload => async dispatch => {
    dispatch({ type: "FETCH_CONTACTS_ERROR", payload });
};