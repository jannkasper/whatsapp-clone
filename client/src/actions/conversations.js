import { publicFetch } from "../util/fetcher";
import { prepare as prepareMessage } from "../util/message"
import { getSocket } from "../util/socket";

export const fetchConversationsPending = () => ({ type: "FETCH_CONVERSATIONS_PENDING" });

export const fetchConversationsSuccess = payload => async dispatch => {
    dispatch({ type: "FETCH_CONVERSATIONS_SUCCESS", payload });
};

export const fetchConversationsError = payload => async dispatch => {
    dispatch({ type: "FETCH_CONVERSATIONS_ERROR", payload });
};

export const selectConversation = payload => async dispatch => {
    dispatch({ type: "SELECT_CONVERSATION", payload });
};

export const createMessage = payload => async (dispatch, getState) => {
    const state = getState();
    const message = await prepareMessage(payload, state);

    dispatch({ type: "CREATE_MESSAGE", payload: { message: message.original }});

    getSocket().emit("MESSAGE", message.toSend);

    // publicFetch.post("message", message.toSend)
    //     .then(response => {
    //         if (!state.conversation.selectedConversation.sessionExtId) {
    //             dispatch({ type: "ENTER_SESSION_IDENTIFIER", payload: {sessionExtId: response.data.sessionExtId }});
    //         }
    //     });
}