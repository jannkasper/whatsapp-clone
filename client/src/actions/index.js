export * from "./app";
export * from "./contacts";
export * from "./messages";