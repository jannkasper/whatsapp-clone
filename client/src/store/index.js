import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import reducers from "../reducers";

const enabledMiddlewares = [thunk];

const middlewares = applyMiddleware(...enabledMiddlewares);

const configureStore = (preloadedState) => {
    const store = createStore(reducers, preloadedState, compose(middlewares));

    return store;
}

export default configureStore