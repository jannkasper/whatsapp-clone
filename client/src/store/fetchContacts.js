import { fetchContactsPending, fetchContactsSuccess, fetchContactsError } from "../actions";
import {publicFetch} from "../util/fetcher";

function fetchContacts() {
    return dispatch => {
        dispatch(fetchContactsPending());
        publicFetch.get("users")
            .then(res => {
                if(res.error) {
                    dispatch(fetchContactsError(res.error));
                }
                dispatch(fetchContactsSuccess({contacts: res.data }));
                return;

            })
            .catch(error => {
                dispatch(fetchContactsError(error));
            })
    }
}

export default fetchContacts