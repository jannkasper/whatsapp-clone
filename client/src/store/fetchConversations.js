import { fetchConversationsPending, fetchConversationsSuccess, fetchConversationsError } from "../actions"
import { publicFetch } from "../util/fetcher";

function fetchConversations(userExtId) {
    return dispatch => {
        dispatch(fetchConversationsPending());
        publicFetch.get(`messages/${userExtId}`)
            .then(res => {
                if(res.error) {
                    dispatch(fetchConversationsError(res.error));
                }
                dispatch(fetchConversationsSuccess({conversations: res.data }));
                return;
            })
            .catch(error => {
                dispatch(fetchConversationsError(error));
            })
    }
}

export default fetchConversations