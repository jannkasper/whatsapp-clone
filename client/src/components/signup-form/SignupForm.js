import React, { useState } from "react";
import { Formik } from "formik";
// import * as Yup from "yup";
import { publicFetch } from "../../util/fetcher"


import profileImage from "../../img/avatar.svg";
import styles from "./signup-form.module.scss";
import fetchContacts from "../../store/fetchContacts";
import fetchConversations from "../../store/fetchConversations";


const SignupForm = ({ handleChangeMode, receiveAuthentication, fetchContacts, fetchConversations }) => {

    const [userImage, setUserImage] = useState(profileImage);
    return (
        <Formik
            initialValues={{ file: null, phoneNumber: "", username: "", password: "" }}
            onSubmit={async (values, { setStatus, resetForm}) => {
                try {
                    let formData = new FormData();
                    formData.append('profileImage', values.file);
                    formData.append('phoneNumber', values.phoneNumber);
                    formData.append('username', values.username);
                    formData.append('password', values.password);

                    const { data } = await publicFetch.post("signup", formData);
                    receiveAuthentication(data);
                    fetchContacts();
                    fetchConversations(data.userInfo.externalIdentifier)
                    resetForm({})
                    setUserImage(profileImage);

                } catch (error) {
                    setStatus(error.response.data.message)
                    console.log(error)
                }
            }}
            // validationSchema={Yup.object({
            //     phoneNumber: Yup.string()
            //         .required("Required")
            //         .length(9, "Must be 9 characters long")
            //         .matches(/^[0-9]+$/, "Contains invalid characters"),
            //     username: Yup.string()
            //         .required("Required")
            //         .max(16, "Must be at most 16 characters long")
            //         .matches(/^[a-zA-Z0-9_-]+$/, "Contains invalid characters"),
            //     password: Yup.string()
            //         .required("Required")
            //         .min(6, "Must be at least 6 characters long")
            //         .max(50, "Must be at most 50 characters long"),
            // })}
        >
            {
                ({
                     values, errors, touched, status, setFieldValue,
                     handleChange, handleBlur, handleSubmit, isSubmitting
                 }) => (
                    <form onSubmit={handleSubmit} className={styles.popup_inner}>
                        <label htmlFor="image">
                            <input
                                id="image"
                                type="file"
                                label="userImage"
                                name="file"
                                accept="image/x-png,image/gif,image/jpeg"
                                value={values.userImage}
                                onChange={(event) => {
                                    setFieldValue("file", event.currentTarget.files[0]);
                                    setUserImage(URL.createObjectURL(event.currentTarget.files[0]))
                                }}
                                onBlur={handleBlur}
                                style={{display: "none"}}
                            />
                            <p className={styles.imageInput}>Edit</p>
                            <img className={styles.image} src={userImage} alt="profileImage"/>
                        </label>
                        <input
                            className={styles.userInput}
                            type="text"
                            name="phoneNumber"
                            autoComplete="off"
                            placeholder="Phone number"
                            value={values.phoneNumber}
                            onChange={handleChange}
                            onBlur={handleBlur}
                        />
                        <input
                            className={styles.userInput}
                            type="text"
                            name="username"
                            autoComplete="off"
                            placeholder="Username"
                            value={values.username}
                            onChange={handleChange}
                            onBlur={handleBlur}
                        />
                        <input
                            className={styles.userInput}
                            type="password"
                            name="password"
                            autoComplete="off"
                            placeholder="Password"
                            value={values.password}
                            onChange={handleChange}
                            onBlur={handleBlur}
                        />
                        <button className={styles.signupButton} type="submit">Sign up</button>
                        <div className={styles.line}></div>
                        <div className={styles.text}>Have an account? <a href="#" onClick={() => handleChangeMode()}>Log in</a></div>
                    </form>
                )
            }
        </Formik>
    )
}

export default SignupForm