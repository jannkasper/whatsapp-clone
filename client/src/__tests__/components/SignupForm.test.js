import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from '../../store';
import SignupForm  from '../../components/signup-form';
import * as actions from '../../actions';
import * as fetchContacts from '../../store/fetchContacts';
import * as fetchConversations from '../../store/fetchConversations';
import * as fetcher from "../../util/fetcher";

const store = configureStore();

jest.mock('../../actions');
jest.mock('../../store/fetchContacts');
jest.mock('../../store/fetchConversations');


describe('SignupForm component', () => {
    it('should display', () => {
        const {asFragment} = render(
            <Provider store={store}>
                <SignupForm />
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    })

    it('can change mode', () => {
        const handleChangeModeMock = jest.fn();
        render(
            <Provider store={store}>
                <SignupForm handleChangeMode={handleChangeModeMock} />
            </Provider>,
        );

        const button = screen.getByText('Log in');

        fireEvent.click(button)

        expect(handleChangeModeMock).toHaveBeenCalled();
    });

    it('can sign up', async () => {
        const data = { userInfo: { externalIdentifier: '1' } };
        fetcher.publicFetch = {
            post: jest.fn(() => Promise.resolve({data})),
        };

        const receiveAuthenticationMock = jest.fn(data => ({type: '', result: data}));
        const fetchContactsMock = jest.fn(() => ({type: ''}));
        const fetchConversationsMock= jest.fn(userExtId => ({type: '', result: userExtId}));

        jest.spyOn(actions, "receiveAuthentication").mockImplementation(receiveAuthenticationMock)
        jest.spyOn(fetchContacts, "default").mockImplementation(fetchContactsMock);
        jest.spyOn(fetchConversations, "default").mockImplementation(fetchConversationsMock)

        render(
            <Provider store={store}>
                <SignupForm />
            </Provider>,
        );

        const phoneNumberInput = screen.getByPlaceholderText('Phone number');
        const usernameInput = screen.getByPlaceholderText('Username');
        const passwordInput = screen.getByPlaceholderText('Password');
        const button = screen.getByText('Sign up');

        fireEvent.change(phoneNumberInput, { target: { value: '555666777' } })
        fireEvent.change(usernameInput, { target: { value: 'usernameTest' } })
        fireEvent.change(passwordInput, { target: { value: 'passwordTest' } })
        fireEvent.click(button)

        await waitFor(() =>  {
            expect(fetcher.publicFetch.post).toHaveBeenCalledTimes(1);
            expect(fetchContactsMock).toHaveBeenLastCalledWith();
            expect(fetchConversationsMock).toHaveBeenLastCalledWith(data.userInfo.externalIdentifier);

        });
    })
});
