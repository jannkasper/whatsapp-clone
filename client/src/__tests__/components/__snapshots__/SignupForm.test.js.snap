// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm component should display 1`] = `
<DocumentFragment>
  <form
    class="popup_inner"
  >
    <label
      for="image"
    >
      <input
        accept="image/x-png,image/gif,image/jpeg"
        id="image"
        label="userImage"
        name="file"
        style="display: none;"
        type="file"
        value=""
      />
      <p
        class="imageInput"
      >
        Edit
      </p>
      <img
        alt="profileImage"
        class="image"
        src="avatar.svg"
      />
    </label>
    <div
      class="container"
    >
      <div
        class="inputContainer"
      >
        <input
          autocomplete="off"
          class="input false"
          for="PhoneNumber"
          name="phoneNumber"
          placeholder="Phone number"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="inputContainer"
      >
        <input
          autocomplete="off"
          class="input false"
          for="Username"
          name="username"
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="inputContainer"
      >
        <input
          autocomplete="off"
          class="input false"
          for="Password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <br />
    <button
      class="signupButton"
      type="submit"
    >
      Sign up
    </button>
    <div
      class="line"
    />
    <div
      class="text"
    >
      Have an account? 
      <a
        href="!#"
      >
        Log in
      </a>
    </div>
  </form>
</DocumentFragment>
`;
